src/main/java/com/redhat/pantheon/model/api/SlingResource.java

```diff
@@ -0,0 +1,274 @@
+package com.redhat.pantheon.model.api;
+
+import org.apache.sling.api.adapter.Adaptable;
+import org.apache.sling.api.resource.*;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.annotation.Nonnull;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.Arrays.stream;
+
+/**
+ * An Editable Sling resource model object. These are simple objects made up of a single constructor based on an existing
+ * sling resource, and a set of fields or child resources mapped to the resource. Fields and child resources are
+ * strongly typed and they can be modified. SlingResources are wrappers around existing sling resources and as such the
+ * resource needs to exist before it can be wrapped by one of these.
+ */
+public class SlingResource implements Adaptable {
+
+    private final Resource resource;
+
+    public SlingResource(@Nonnull Resource resource) {
+        this.resource = resource;
+    }
+
+    public Resource getResource() {
+        return resource;
+    }
+
+    /**
+     * Initializes all fields which have a default value set, and which aren't already set.
+     * This method is only meant to be called after new resources are created.
+     */
+    protected void initDefaultValues() {
+        getMembers(Field.class)
+                // discard fields which already have a value set
+                .filter(field -> !field.isSet())
+                // only consider fields which have a default value
+                .filter(field -> field.defaultValue.isPresent())
+                .forEach(field -> field.set(field.defaultValue.get()));
+    }
+
+    /**
+     * Returns all SlingModel members (assigned fields which implement the ResourceMember interdace
+     * @return
+     */
+    private Stream<ResourceMember> allMembers() {
+        return stream(this.getClass().getDeclaredFields())
+                // only fields which implement ResourceMember
+                .filter(reflectedField -> ResourceMember.class.isAssignableFrom(reflectedField.getType()))
+                // convert to the field values
+                .map(field -> {
+                    try {
+                        return (ResourceMember) field.get(SlingResource.this);
+                    } catch (IllegalAccessException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                // only initialized fields
+                .filter(field -> field != null);
```

Comment:
You are entirely correct in your analysis of the code (I've made a few comment changes to reflect your wording). And yes, the null check is preventing fields which are never initialized (or initialized to null), since there is nothing the API will ever need to do with these fields.
I can change the name of the method to something like `allInitializedMembers` to better reflect what it's doing.

Reply:
It's a private method, so the naming convention is less important. It looks like you've documented it nicely with comments, so that's sufficient. Thanks!

I'm going to assume that for my last question regarding `Field whatever = null` is not a valid scenario. I think the developer would be doing something really weird and probably wrong, but either way it doesn't hurt to leave the null check in place.

Reply:
:+1: